### ISSUE TITLE
[Question] Why should use both RabbitMQ & Redis ?

### ISSUE THREAD

**Issue by nvcken** [poanetwork/bridge-nodejs#90]
Hi, anyone pls help to explain, thanks

**Comment by akolotov** [comment]
Hi,
We have several workers in bridge-nodejs: 3 to watch for events and 2 to send transactions.
RabbitMQ is used to send jobs from watchers to senders.
Raddis is used to store number of blocks that were already inspected by watchers, and the NOnce which was used by the sender last time to send a transaction (it is done if we want to scale senders in the future).

**Comment by nvcken** [comment]
Nice, just 2 more questions
1. Can Redis use to do the same tasks likes RabbitMQ ( send jobs from watchers to senders ), what pros take from RabbitMQ over Redis for this tasks?
2. how duplicate Nonce issue when send raw transaction and solves by store that value on Redis ?

**Comment by fvictorio** [comment]
> Can Redis use to do the same tasks likes RabbitMQ ( send jobs from watchers to senders ), what pros take from RabbitMQ over Redis for this tasks?

There are libraries for node that implement job queues on top of Redis, but none of the ones I know is as powerful and stable as rabbit.

> 2\. how duplicate Nonce issue when send raw transaction and solves by store that value on Redis ?

I didn't understand this question, could you rephrase it?

**Comment by nvcken** [comment]
sorry, the second about Nonce value store on redis, does it purpose to prevent duplicate Nonce value on more than one of asynchronous sending transactions

**Comment by fvictorio** [comment]
We keep track of the nonce value to be able to send multiple transactions at the same time (with nonces N, N+1, N+2, etc.). The challenge is to avoid creating transactions that have the same nonce, or that make "gaps" in the nonce values (for example: if the account made 10 transactions and you create one with nonce 12, it won't be processed until a transaction with nonce 11 is mined).

**Comment by nvcken** [comment]
Many thanks for you guys.
